File: java-sdk/src/main/java/io/graphlite/sdk/ffi/GraphLiteFFI.java
package io.graphlite.sdk.ffi;

import io.graphlite.sdk.Errors;
import io.graphlite.sdk.Errors.ErrorCode;

import java.lang.foreign.*;
import java.lang.invoke.MethodHandle;
import java.nio.file.Files;
import java.nio.file.Path;

/**
 * Centralizes all Foreign Function &amp; Memory API (Panama FFM) bindings to the
 * {@code libgraphlite_ffi} shared library.
 * <p>
 * Every native call is a thin wrapper around a {@link MethodHandle} downcall
 * obtained from {@link Linker#downcallHandle}.  Callers should prefer the
 * high-level {@link io.graphlite.sdk.GraphLite} / {@link io.graphlite.sdk.Session}
 * classes instead of invoking these handles directly.
 * <p>
 * <b>Library resolution order:</b>
 * <ol>
 *   <li>Environment variable {@code GRAPHLITE_FFI_LIB} &ndash; full path to the shared library.</li>
 *   <li>System property {@code graphlite.ffi.lib} &ndash; full path to the shared library.</li>
 *   <li>Platform-default search paths (respects {@code LD_LIBRARY_PATH} / {@code DYLD_LIBRARY_PATH} / {@code PATH}).</li>
 * </ol>
 */
public final class GraphLiteFFI {

    private GraphLiteFFI() {}

    private static final Linker LINKER = Linker.nativeLinker();

    // --- Method handles (initialized lazily on first access) ---

    private static volatile SymbolLookup lookup;
    private static volatile Arena sharedArena;

    private static MethodHandle MH_open;
    private static MethodHandle MH_createSession;
    private static MethodHandle MH_query;
    private static MethodHandle MH_closeSession;
    private static MethodHandle MH_freeString;
    private static MethodHandle MH_close;
    private static MethodHandle MH_version;

    // --- Public initialization ---

    /**
     * Initialize the FFI layer by loading the native library.
     * This is idempotent; subsequent calls are no-ops.
     *
     * @throws Errors.LibraryLoadException if the library cannot be found/loaded.
     */
    public static synchronized void init() {
        if (lookup != null) return;
        try {
            sharedArena = Arena.ofShared();
            lookup = resolveLibrary(sharedArena);
            bindAll();
        } catch (Errors.LibraryLoadException e) {
            throw e;
        } catch (Exception e) {
            throw new Errors.LibraryLoadException("Failed to initialize GraphLite FFI", e);
        }
    }

    // --- Downcall wrappers -----------------------------------------------

    /**
     * {@code GraphLiteDB* graphlite_open(const char* path, GraphLiteErrorCode* error_out)}
     *
     * @return native handle (opaque pointer) to the database, or {@link MemorySegment#NULL} on error.
     */
    public static MemorySegment open(MemorySegment path, MemorySegment errorOut) {
        try {
            return (MemorySegment) MH_open.invokeExact(path, errorOut);
        } catch (Throwable t) { throw rethrow(t); }
    }

    /**
     * {@code char* graphlite_create_session(GraphLiteDB* db, const char* username, GraphLiteErrorCode* error_out)}
     */
    public static MemorySegment createSession(MemorySegment db, MemorySegment username, MemorySegment errorOut) {
        try {
            return (MemorySegment) MH_createSession.invokeExact(db, username, errorOut);
        } catch (Throwable t) { throw rethrow(t); }
    }

    /**
     * {@code char* graphlite_query(GraphLiteDB* db, const char* session_id, const char* query, GraphLiteErrorCode* error_out)}
     */
    public static MemorySegment query(MemorySegment db, MemorySegment sessionId, MemorySegment query, MemorySegment errorOut) {
        try {
            return (MemorySegment) MH_query.invokeExact(db, sessionId, query, errorOut);
        } catch (Throwable t) { throw rethrow(t); }
    }

    /**
     * {@code GraphLiteErrorCode graphlite_close_session(GraphLiteDB* db, const char* session_id, GraphLiteErrorCode* error_out)}
     */
    public static int closeSession(MemorySegment db, MemorySegment sessionId, MemorySegment errorOut) {
        try {
            return (int) MH_closeSession.invokeExact(db, sessionId, errorOut);
        } catch (Throwable t) { throw rethrow(t); }
    }

    /**
     * {@code void graphlite_free_string(char* s)}
     */
    public static void freeString(MemorySegment s) {
        try {
            MH_freeString.invokeExact(s);
        } catch (Throwable t) { throw rethrow(t); }
    }

    /**
     * {@code void graphlite_close(GraphLiteDB* db)}
     */
    public static void close(MemorySegment db) {
        try {
            MH_close.invokeExact(db);
        } catch (Throwable t) { throw rethrow(t); }
    }

    /**
     * {@code const char* graphlite_version()}
     * Returns a pointer to a static string &mdash; must <b>not</b> be freed.
     */
    public static MemorySegment version() {
        try {
            return (MemorySegment) MH_version.invokeExact();
        } catch (Throwable t) { throw rethrow(t); }
    }

    // --- Utility helpers for callers --------------------------------------

    /** Allocate a null-terminated UTF-8 C string inside the given arena. */
    public static MemorySegment toCString(String s, Arena arena) {
        return arena.allocateUtf8String(s);
    }

    /** Read a null-terminated UTF-8 C string from a native pointer. */
    public static String fromCString(MemorySegment ptr) {
        return ptr.reinterpret(Long.MAX_VALUE).getUtf8String(0);
    }

    /** Allocate a single-int segment for the error-out parameter. */
    public static MemorySegment allocErrorOut(Arena arena) {
        MemorySegment seg = arena.allocate(Layouts.ERROR_CODE);
        seg.set(Layouts.ERROR_CODE, 0, 0);
        return seg;
    }

    /** Read the error code written by the native side. */
    public static ErrorCode readErrorCode(MemorySegment errorOut) {
        return ErrorCode.fromNative(errorOut.get(Layouts.ERROR_CODE, 0));
    }

    // --- Internal ---------------------------------------------------------

    private static SymbolLookup resolveLibrary(Arena arena) {
        String envPath = System.getenv("GRAPHLITE_FFI_LIB");
        if (envPath != null && !envPath.isBlank()) {
            return loadFrom(Path.of(envPath), arena);
        }
        String sysProp = System.getProperty("graphlite.ffi.lib");
        if (sysProp != null && !sysProp.isBlank()) {
            return loadFrom(Path.of(sysProp), arena);
        }

        String libFile = platformLibraryName();
        String[] candidates = {
            "target/release/" + libFile,
            "../target/release/" + libFile,
            "../../target/release/" + libFile,
            "target/debug/" + libFile,
            "../target/debug/" + libFile,
            "../../target/debug/" + libFile,
        };
        for (String candidate : candidates) {
            Path p = Path.of(candidate).toAbsolutePath().normalize();
            if (Files.isRegularFile(p)) {
                return loadFrom(p, arena);
            }
        }

        try {
            return SymbolLookup.libraryLookup(System.mapLibraryName("graphlite_ffi"), arena);
        } catch (IllegalArgumentException e) {
            throw new Errors.LibraryLoadException(
                "Could not locate libgraphlite_ffi. Set GRAPHLITE_FFI_LIB or " +
                "LD_LIBRARY_PATH/DYLD_LIBRARY_PATH to the directory containing the library. " +
                "Build with: cargo build --release -p graphlite-ffi", e);
        }
    }

    private static SymbolLookup loadFrom(Path path, Arena arena) {
        if (!Files.isRegularFile(path)) {
            throw new Errors.LibraryLoadException(
                "Library not found at " + path.toAbsolutePath(), null);
        }
        return SymbolLookup.libraryLookup(path, arena);
    }

    private static String platformLibraryName() {
        String os = System.getProperty("os.name", "").toLowerCase();
        if (os.contains("mac") || os.contains("darwin")) {
            return "libgraphlite_ffi.dylib";
        } else if (os.contains("win")) {
            return "graphlite_ffi.dll";
        }
        return "libgraphlite_ffi.so";
    }

    private static MemorySegment findSymbol(String name) {
        return lookup.find(name).orElseThrow(() ->
            new Errors.LibraryLoadException("Symbol not found: " + name, null));
    }

    private static void bindAll() {
        // GraphLiteDB* graphlite_open(const char*, GraphLiteErrorCode*)
        MH_open = LINKER.downcallHandle(
            findSymbol("graphlite_open"),
            FunctionDescriptor.of(ValueLayout.ADDRESS, ValueLayout.ADDRESS, ValueLayout.ADDRESS)
        );
        // char* graphlite_create_session(GraphLiteDB*, const char*, GraphLiteErrorCode*)
        MH_createSession = LINKER.downcallHandle(
            findSymbol("graphlite_create_session"),
            FunctionDescriptor.of(ValueLayout.ADDRESS, ValueLayout.ADDRESS, ValueLayout.ADDRESS, ValueLayout.ADDRESS)
        );
        // char* graphlite_query(GraphLiteDB*, const char*, const char*, GraphLiteErrorCode*)
        MH_query = LINKER.downcallHandle(
            findSymbol("graphlite_query"),
            FunctionDescriptor.of(ValueLayout.ADDRESS, ValueLayout.ADDRESS, ValueLayout.ADDRESS, ValueLayout.ADDRESS, ValueLayout.ADDRESS)
        );
        // GraphLiteErrorCode graphlite_close_session(GraphLiteDB*, const char*, GraphLiteErrorCode*)
        MH_closeSession = LINKER.downcallHandle(
            findSymbol("graphlite_close_session"),
            FunctionDescriptor.of(ValueLayout.JAVA_INT, ValueLayout.ADDRESS, ValueLayout.ADDRESS, ValueLayout.ADDRESS)
        );
        // void graphlite_free_string(char*)
        MH_freeString = LINKER.downcallHandle(
            findSymbol("graphlite_free_string"),
            FunctionDescriptor.ofVoid(ValueLayout.ADDRESS)
        );
        // void graphlite_close(GraphLiteDB*)
        MH_close = LINKER.downcallHandle(
            findSymbol("graphlite_close"),
            FunctionDescriptor.ofVoid(ValueLayout.ADDRESS)
        );
        // const char* graphlite_version()
        MH_version = LINKER.downcallHandle(
            findSymbol("graphlite_version"),
            FunctionDescriptor.of(ValueLayout.ADDRESS)
        );
    }

    @SuppressWarnings("unchecked")
    private static RuntimeException rethrow(Throwable t) {
        if (t instanceof RuntimeException re) throw re;
        if (t instanceof Error e) throw e;
        throw new RuntimeException(t);
    }
}

File: lua-sdk/src/graphlite_ffi.lua
--[[
  GraphLite FFI - Thin adapter over LuaJIT ffi for the GraphLite C FFI library.
  Binds against libgraphlite_ffi (same symbols as Python/Java bindings).
  All ffi.cdef and ffi.load logic lives here.
]]

local ffi = require("ffi")

-- Platform-aware library loading
local function find_library()
  local lib_name
  if ffi.os == "OSX" then
    lib_name = "libgraphlite_ffi.dylib"
  elseif ffi.os == "Windows" then
    lib_name = "graphlite_ffi.dll"
  else
    lib_name = "libgraphlite_ffi.so"
  end

  -- Infer GraphLite repo root from this file's location (lua-sdk/src/graphlite_ffi.lua)
  -- Repo root is parent of lua-sdk, so target/ is at repo_root/target/
  local repo_root
  local src = debug.getinfo(1, "S").source
  if src:sub(1, 1) == "@" then
    src = src:sub(2)
  end
  local src_dir = src:match("(.+)/[^/]+$") or "."
  if src_dir:match("/src$") then
    local lua_sdk_root = src_dir:match("(.+)/src$")
    if lua_sdk_root and lua_sdk_root:match("/lua%-sdk$") then
      repo_root = lua_sdk_root:match("(.+)/lua%-sdk$")
    elseif lua_sdk_root then
      repo_root = lua_sdk_root  -- lua-sdk at repo root
    end
  end
  if repo_root and repo_root ~= "" then
    repo_root = repo_root:gsub("/+$", "")
  end

  local search_paths = {}
  if repo_root then
    search_paths[#search_paths + 1] = repo_root .. "/target/release/" .. lib_name
    search_paths[#search_paths + 1] = repo_root .. "/target/debug/" .. lib_name
  end
  for _, p in ipairs({
    "graphlite_ffi", "libgraphlite_ffi", lib_name,
    "target/release/" .. lib_name, "target/debug/" .. lib_name,
    "../target/release/" .. lib_name, "../target/debug/" .. lib_name,
    "../../target/release/" .. lib_name, "../../target/debug/" .. lib_name,
    "../../../target/release/" .. lib_name, "../../../target/debug/" .. lib_name,
    "/usr/local/lib/" .. lib_name, "/usr/lib/" .. lib_name,
  }) do
    search_paths[#search_paths + 1] = p
  end

  for _, path in ipairs(search_paths) do
    local ok, lib = pcall(ffi.load, path)
    if ok and lib then
      return lib
    end
  end

  return nil, "Could not find GraphLite library (" .. lib_name .. "). " ..
    "Build first: cargo build --release -p graphlite-ffi"
end

-- C FFI definitions (match graphlite-ffi/src/lib.rs)
ffi.cdef([[
  typedef enum {
    GraphLiteErrorCode_Success = 0,
    GraphLiteErrorCode_NullPointer = 1,
    GraphLiteErrorCode_InvalidUtf8 = 2,
    GraphLiteErrorCode_DatabaseOpenError = 3,
    GraphLiteErrorCode_SessionError = 4,
    GraphLiteErrorCode_QueryError = 5,
    GraphLiteErrorCode_PanicError = 6,
    GraphLiteErrorCode_JsonError = 7
  } GraphLiteErrorCode;

  typedef struct GraphLiteDB GraphLiteDB;

  GraphLiteDB* graphlite_open(const char* path, GraphLiteErrorCode* error_out);
  char* graphlite_create_session(GraphLiteDB* db, const char* username, GraphLiteErrorCode* error_out);
  char* graphlite_query(GraphLiteDB* db, const char* session_id, const char* query, GraphLiteErrorCode* error_out);
  GraphLiteErrorCode graphlite_close_session(GraphLiteDB* db, const char* session_id, GraphLiteErrorCode* error_out);
  void graphlite_free_string(char* s);
  void graphlite_close(GraphLiteDB* db);
  const char* graphlite_version(void);
]])

local lib, err = find_library()
if not lib then
  error(err)
end

-- Error code names for Lua
local ERROR_NAMES = {
  [0] = "Success",
  [1] = "NullPointer",
  [2] = "InvalidUtf8",
  [3] = "DatabaseOpenError",
  [4] = "SessionError",
  [5] = "QueryError",
  [6] = "PanicError",
  [7] = "JsonError",
}

local M = {}

function M.open(path)
  local error_out = ffi.new("GraphLiteErrorCode[1]")
  error_out[0] = 0
  local db = lib.graphlite_open(path, error_out)
  if db == nil then
    return nil, error_out[0], ERROR_NAMES[error_out[0]] or "Unknown"
  end
  return db
end

function M.create_session(db, username)
  local error_out = ffi.new("GraphLiteErrorCode[1]")
  error_out[0] = 0
  local session_id_ptr = lib.graphlite_create_session(db, username, error_out)
  if session_id_ptr == nil then
    return nil, error_out[0], ERROR_NAMES[error_out[0]] or "Unknown"
  end
  local session_id = ffi.string(session_id_ptr)
  lib.graphlite_free_string(session_id_ptr)
  return session_id
end

function M.query(db, session_id, query_str)
  local error_out = ffi.new("GraphLiteErrorCode[1]")
  error_out[0] = 0
  local result_ptr = lib.graphlite_query(db, session_id, query_str, error_out)
  if result_ptr == nil then
    return nil, error_out[0], ERROR_NAMES[error_out[0]] or "Unknown"
  end
  local json_str = ffi.string(result_ptr)
  lib.graphlite_free_string(result_ptr)
  return json_str
end

function M.close_session(db, session_id)
  local error_out = ffi.new("GraphLiteErrorCode[1]")
  error_out[0] = 0
  local code = lib.graphlite_close_session(db, session_id, error_out)
  return code == 0, error_out[0]
end

function M.close(db)
  if db and db ~= nil then
    lib.graphlite_close(db)
  end
end

function M.version()
  local v = lib.graphlite_version()
  if v and v ~= nil then
    return ffi.string(v)
  end
  return "unknown"
end

M.ERROR_NAMES = ERROR_NAMES
M.lib = lib

return M
